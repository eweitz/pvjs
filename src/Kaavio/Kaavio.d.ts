/// <reference types="react" />
import * as React from 'react';
/**
 * Kaavio component.
 * This is the highest component in Kaavio. All states are handled here and passed down as props to other components.
 *
 * You may pass an onReady(kaavio) function to this. This will be called with the Kaavio reference when everything is
 * rendered. You can access the manipulation API via kaavio.manipulator
 */
export declare class Kaavio extends React.Component<any, any> {
    constructor(props: any);
    onPanZoomReady: (panZoom: any) => void;
    handleClick: (e: any) => void;
    render(): JSX.Element;
}
