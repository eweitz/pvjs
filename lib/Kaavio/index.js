"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Kaavio_1 = require("./Kaavio");
exports.Kaavio = Kaavio_1.Kaavio;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi9zcmMvS2FhdmlvL2luZGV4LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBQUEsbUNBQWdDO0FBQXhCLDBCQUFBLE1BQU0sQ0FBQSJ9