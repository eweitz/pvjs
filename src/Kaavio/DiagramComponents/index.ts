import {Edge} from "./Edge";
import {Node} from './Node'
import {Marker} from './Marker';
import {Group} from './Group';
import {NonFuncIriMarkerPropertyValue, MarkerPropertyName} from "./typings";
export * from './typings.d.ts';

export const Components = {
    Edge: Edge,
    Node: Node,
    Marker: Marker,
    Group: Group
};

export const NON_FUNC_IRI_MARKER_PROPERTY_VALUES: NonFuncIriMarkerPropertyValue[] = ['none', 'inherit'];
export const MARKER_PROPERTY_NAMES: MarkerPropertyName[] = ['markerStart', 'markerMid', 'markerEnd', 'marker'];