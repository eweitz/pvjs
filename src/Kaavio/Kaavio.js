"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var React = require("react");
var Diagram_1 = require("./components/Diagram");
var PanZoom_1 = require("./components/PanZoom");
/**
 * Kaavio component.
 * This is the highest component in Kaavio. All states are handled here and passed down as props to other components.
 *
 * You may pass an onReady(kaavio) function to this. This will be called with the Kaavio reference when everything is
 * rendered. You can access the manipulation API via kaavio.manipulator
 */
var Kaavio = (function (_super) {
    __extends(Kaavio, _super);
    function Kaavio(props) {
        var _this = _super.call(this, props) || this;
        _this.onPanZoomReady = function (panZoom) {
            // Fire the onReady function with a reference to Kaavio
            var onReady = _this.props.onReady;
            onReady(_this);
        };
        _this.handleClick = function (e) {
            var onEntityClick = _this.props.onEntityClick;
            var entity = e.entity;
            if (onEntityClick && entity)
                onEntityClick(entity);
        };
        _this.state = {
            diagramRef: null,
        };
        return _this;
    }
    Kaavio.prototype.render = function () {
        var _this = this;
        var _a = this.props, customStyle = _a.customStyle, filters = _a.filters, entities = _a.entities, name = _a.name, width = _a.width, height = _a.height, edgeDrawers = _a.edgeDrawers, icons = _a.icons, markerDrawers = _a.markerDrawers, highlightedEntities = _a.highlightedEntities, hiddenEntities = _a.hiddenEntities, zoomedEntities = _a.zoomedEntities, pannedEntities = _a.pannedEntities, _b = _a.showPanZoomControls, showPanZoomControls = _b === void 0 ? true : _b;
        var backgroundColor = customStyle.backgroundColor || 'white';
        var about = 'kaavio-container';
        // This is a port to the legacy use of node_id
        // TODO: Change all references of highlightedNodes to the new highlightedEntities
        var highlightedEntitiesLegacy = highlightedEntities ? highlightedEntities.map(function (singleHighlightedEntity) {
            return {
                node_id: singleHighlightedEntity.entityId,
                color: singleHighlightedEntity.color
            };
        }) : [];
        var entityMap = entities.reduce(function (acc, entity) {
            acc[entity.id] = entity;
            return acc;
        }, {});
        var zIndices = entities
            .sort(function (a, b) {
            if (a.zIndex > b.zIndex) {
                return 1;
            }
            else if (a.zIndex < b.zIndex) {
                return -1;
            }
            else {
                return 0;
            }
        })
            .map(function (entity) { return entity.id; });
        // TODO: Don't use refs!
        // Accessing the diagram ref from the state is a little bit of a hack to get panZoom working.
        // Consider refactoring the panZoom to be truly Reactive and not use refs
        return (React.createElement("div", { id: about, className: "kaavio-container " + customStyle.containerClass },
            React.createElement(Diagram_1.Diagram, { ref: function (diagram) { return !_this.state.diagramRef && _this.setState({ diagramRef: diagram }); }, about: "kaavio-diagram-for-" + about, name: name, width: width, height: height, backgroundColor: backgroundColor, edgeDrawers: edgeDrawers, entityMap: entityMap, filters: filters, handleClick: this.handleClick, icons: icons, markerDrawers: markerDrawers, zIndices: zIndices, customStyle: customStyle, highlightedNodes: highlightedEntitiesLegacy, hiddenEntities: hiddenEntities }),
            React.createElement(PanZoom_1.PanZoom, { diagram: this.state.diagramRef, zoomedEntities: zoomedEntities, pannedEntities: pannedEntities, onReady: this.onPanZoomReady, showPanZoomControls: showPanZoomControls })));
    };
    return Kaavio;
}(React.Component));
exports.Kaavio = Kaavio;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiS2FhdmlvLmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiS2FhdmlvLnRzeCJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOzs7Ozs7Ozs7OztBQUFBLDZCQUErQjtBQUMvQixnREFBNkM7QUFDN0MsZ0RBQTZDO0FBRTdDOzs7Ozs7R0FNRztBQUNIO0lBQTRCLDBCQUF5QjtJQUVwRCxnQkFBWSxLQUFLO1FBQWpCLFlBQ0Msa0JBQU0sS0FBSyxDQUFDLFNBSVo7UUFFRCxvQkFBYyxHQUFHLFVBQUMsT0FBTztZQUN4Qix1REFBdUQ7WUFDaEQsSUFBQSw2QkFBTyxDQUFlO1lBQzdCLE9BQU8sQ0FBQyxLQUFJLENBQUMsQ0FBQztRQUNmLENBQUMsQ0FBQztRQUVGLGlCQUFXLEdBQUcsVUFBQyxDQUFDO1lBQ1AsSUFBQSx5Q0FBYSxDQUFnQjtZQUNyQyxJQUFNLE1BQU0sR0FBRyxDQUFDLENBQUMsTUFBTSxDQUFDO1lBQ3hCLEVBQUUsQ0FBQyxDQUFDLGFBQWEsSUFBSSxNQUFNLENBQUM7Z0JBQzNCLGFBQWEsQ0FBQyxNQUFNLENBQUMsQ0FBQztRQUN4QixDQUFDLENBQUE7UUFoQkEsS0FBSSxDQUFDLEtBQUssR0FBRztZQUNaLFVBQVUsRUFBRSxJQUFJO1NBQ2hCLENBQUM7O0lBQ0gsQ0FBQztJQWVELHVCQUFNLEdBQU47UUFBQSxpQkFnRUM7UUEvRE0sSUFBQSxlQUVtQyxFQUZsQyw0QkFBVyxFQUFFLG9CQUFPLEVBQUUsc0JBQVEsRUFBRSxjQUFJLEVBQUUsZ0JBQUssRUFBRSxrQkFBTSxFQUFFLDRCQUFXLEVBQUUsZ0JBQUssRUFDN0UsZ0NBQWEsRUFBRSw0Q0FBbUIsRUFBRSxrQ0FBYyxFQUFFLGtDQUFjLEVBQUUsa0NBQWMsRUFDbEYsMkJBQTBCLEVBQTFCLCtDQUEwQixDQUFlO1FBRTFDLElBQU0sZUFBZSxHQUFHLFdBQVcsQ0FBQyxlQUFlLElBQUksT0FBTyxDQUFFO1FBQ2hFLElBQU0sS0FBSyxHQUFHLGtCQUFrQixDQUFDO1FBRWpDLDhDQUE4QztRQUM5QyxpRkFBaUY7UUFDakYsSUFBTSx5QkFBeUIsR0FBRyxtQkFBbUIsR0FBRyxtQkFBbUIsQ0FBQyxHQUFHLENBQUMsVUFBQSx1QkFBdUI7WUFDdEcsTUFBTSxDQUFDO2dCQUNOLE9BQU8sRUFBRSx1QkFBdUIsQ0FBQyxRQUFRO2dCQUN6QyxLQUFLLEVBQUUsdUJBQXVCLENBQUMsS0FBSzthQUNwQyxDQUFBO1FBQ0YsQ0FBQyxDQUFDLEdBQUcsRUFBRSxDQUFDO1FBRVIsSUFBTSxTQUFTLEdBQUcsUUFBUSxDQUFDLE1BQU0sQ0FBQyxVQUFTLEdBQUcsRUFBRSxNQUFNO1lBQ3JELEdBQUcsQ0FBQyxNQUFNLENBQUMsRUFBRSxDQUFDLEdBQUcsTUFBTSxDQUFDO1lBQ3hCLE1BQU0sQ0FBQyxHQUFHLENBQUM7UUFDWixDQUFDLEVBQUUsRUFBRSxDQUFDLENBQUM7UUFFUCxJQUFNLFFBQVEsR0FBRyxRQUFRO2FBQ3ZCLElBQUksQ0FBQyxVQUFTLENBQU0sRUFBRSxDQUFNO1lBQzVCLEVBQUUsQ0FBQyxDQUFDLENBQUMsQ0FBQyxNQUFNLEdBQUcsQ0FBQyxDQUFDLE1BQU0sQ0FBQyxDQUFDLENBQUM7Z0JBQ3pCLE1BQU0sQ0FBQyxDQUFDLENBQUM7WUFDVixDQUFDO1lBQUMsSUFBSSxDQUFDLEVBQUUsQ0FBQyxDQUFDLENBQUMsQ0FBQyxNQUFNLEdBQUcsQ0FBQyxDQUFDLE1BQU0sQ0FBQyxDQUFDLENBQUM7Z0JBQ2hDLE1BQU0sQ0FBQyxDQUFDLENBQUMsQ0FBQztZQUNYLENBQUM7WUFBQyxJQUFJLENBQUMsQ0FBQztnQkFDUCxNQUFNLENBQUMsQ0FBQyxDQUFDO1lBQ1YsQ0FBQztRQUNGLENBQUMsQ0FBQzthQUNELEdBQUcsQ0FBQyxVQUFDLE1BQU0sSUFBSyxPQUFBLE1BQU0sQ0FBQyxFQUFFLEVBQVQsQ0FBUyxDQUFDLENBQUM7UUFFN0Isd0JBQXdCO1FBQ3hCLDZGQUE2RjtRQUM3Rix5RUFBeUU7UUFDekUsTUFBTSxDQUFDLENBQ04sNkJBQUssRUFBRSxFQUFFLEtBQUssRUFBRSxTQUFTLEVBQUUsc0JBQXFCLFdBQVcsQ0FBQyxjQUFpQjtZQUM1RSxvQkFBQyxpQkFBTyxJQUNQLEdBQUcsRUFBRSxVQUFBLE9BQU8sSUFBSSxPQUFBLENBQUMsS0FBSSxDQUFDLEtBQUssQ0FBQyxVQUFVLElBQUksS0FBSSxDQUFDLFFBQVEsQ0FBQyxFQUFDLFVBQVUsRUFBRSxPQUFPLEVBQUMsQ0FBQyxFQUE5RCxDQUE4RCxFQUM5RSxLQUFLLEVBQUUsd0JBQXNCLEtBQU8sRUFDcEMsSUFBSSxFQUFFLElBQUksRUFDVixLQUFLLEVBQUUsS0FBSyxFQUNaLE1BQU0sRUFBRSxNQUFNLEVBQ2QsZUFBZSxFQUFFLGVBQWUsRUFDaEMsV0FBVyxFQUFFLFdBQVcsRUFDeEIsU0FBUyxFQUFFLFNBQVMsRUFDcEIsT0FBTyxFQUFFLE9BQU8sRUFDaEIsV0FBVyxFQUFFLElBQUksQ0FBQyxXQUFXLEVBQzdCLEtBQUssRUFBRSxLQUFLLEVBQ1osYUFBYSxFQUFFLGFBQWEsRUFDNUIsUUFBUSxFQUFFLFFBQVEsRUFDbEIsV0FBVyxFQUFFLFdBQVcsRUFDeEIsZ0JBQWdCLEVBQUUseUJBQXlCLEVBQzNDLGNBQWMsRUFBRSxjQUFjLEdBQzdCO1lBQ0Ysb0JBQUMsaUJBQU8sSUFBQyxPQUFPLEVBQUUsSUFBSSxDQUFDLEtBQUssQ0FBQyxVQUFVLEVBQ3BDLGNBQWMsRUFBRSxjQUFjLEVBQzlCLGNBQWMsRUFBRSxjQUFjLEVBQzlCLE9BQU8sRUFBRSxJQUFJLENBQUMsY0FBYyxFQUM1QixtQkFBbUIsRUFBRSxtQkFBbUIsR0FBSSxDQUMxQyxDQUNOLENBQUE7SUFDRixDQUFDO0lBQ0YsYUFBQztBQUFELENBQUMsQUF2RkQsQ0FBNEIsS0FBSyxDQUFDLFNBQVMsR0F1RjFDO0FBdkZZLHdCQUFNIn0=