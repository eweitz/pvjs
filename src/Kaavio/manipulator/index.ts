import {Highlighter} from './highlighter';
import {Hider} from "./hider";
import {Zoomer} from "./zoomer";
import {Panner} from "./panner";
import * as ReactDOM from 'react-dom';

export function getManipulator(kaavioRef, panZoomRef, diagramRef) {
    const highlighter = new Highlighter(kaavioRef);
    const hider = new Hider(kaavioRef);
    const diagram = ReactDOM.findDOMNode(diagramRef);
    const zoomer = new Zoomer(kaavioRef, panZoomRef, diagram);
    const panner = new Panner(kaavioRef, panZoomRef, diagram);

    return {
        toggleHighlight: highlighter.toggleHighlight.bind(highlighter),
        highlightOn: highlighter.highlightOn.bind(highlighter),
        highlightOff: highlighter.highlightOff.bind(highlighter),
        resetHighlighted: highlighter.resetHighlighted.bind(highlighter),
        hide: hider.hide.bind(hider),
        show: hider.show.bind(hider),
        resetHidden: hider.resetHidden.bind(hider),
        zoomOn: function(entity_id: string | string[]) {
            panner.panTo.bind(panner)(entity_id).then(() => {
                zoomer.zoomOn.bind(zoomer)(entity_id);
            });
        },
        zoomIn: zoomer.zoomIn.bind(zoomer),
        zoomOut: zoomer.zoomOut.bind(zoomer),
        zoom: zoomer.zoom.bind(zoomer),
        resetZoom: zoomer.resetZoom.bind(zoomer),
        panTo: panner.panTo.bind(panner),
        pan: panner.pan.bind(panner),
        resetPan: panner.resetPan.bind(panner),
        reset: function () {
            this.resetPan();
            this.resetZoom();
            this.resetHidden();
            this.resetHighlighted();
        },
        getEntities: function () {
            return kaavioRef.getEntities().map(entity => {
                return {
                    id: entity.id,
                    kaavioType: entity.kaavioType,
                    textContent: entity.textContent,
                    types: entity.type
                }
            });
        }
    }
}
