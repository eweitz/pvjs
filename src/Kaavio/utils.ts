import {
    MarkerPropertyName, NonFuncIriMarkerPropertyValue,
    NON_FUNC_IRI_MARKER_PROPERTY_VALUES
} from "./DiagramComponents";
import {highlightedNode} from "./Kaavio";
import * as _ from 'lodash';

export function getMarkerId(
    markerLocationType: MarkerPropertyName,
    markerName: string,
    color: string,
    backgroundColor: string
): string {
    return [markerLocationType, markerName, color, backgroundColor]
        .join('-')
        // we only want alphanumeric values and dashes in the id
        .replace(/[^A-Za-z0-9-]/g, '');
}

export function getMarkerPropertyValue(
    markerLocationType: MarkerPropertyName,
    markerName: NonFuncIriMarkerPropertyValue & string,
    color: string,
    backgroundColor: string
): NonFuncIriMarkerPropertyValue | string {
    // Don't make a funciri out of any of the names in NON_FUNC_IRI_MARKER_PROPERTY_VALUES
    if (NON_FUNC_IRI_MARKER_PROPERTY_VALUES.indexOf(markerName) > -1) {
        return markerName;
    }
    return `url(#${getMarkerId(markerLocationType, markerName, color, backgroundColor)})`;
}

export function getHighlighted(entity, highlightedNodes) {
    let result = {
        highlighted: false,
        color: null
    };
    // Only allow nodes to be highlighted
    if(entity.kaavioType != 'Node') return result;

    let matched: highlightedNode = _.find(highlightedNodes, (value, index) => {
        return value['node_id'] == entity.id;
    }) as highlightedNode;

    if (matched) {
        result.highlighted = true;
        result.color = matched.color;
    }

    return result;
}