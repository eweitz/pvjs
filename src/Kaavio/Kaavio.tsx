import {forOwn} from 'lodash';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {Diagram} from './components/Diagram';
// TODO see whether there's anything I need in here. If not, delete.
//require('./kaavio.css');
import {PanZoom} from "./components/PanZoom";
import {Manipulator} from './manipulator';

// pullAllWith is missing from the lodash typings so just require for now
// See issue: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/13747
// TODO: Track the issue and go back to import * as _ from 'lodash'
const _ = require('lodash');

export interface highlightedNode {
	node_id: string,
	color: string; // CSS color. E.g. 'red' or '#ffff'
}

/**
 * Kaavio component.
 * This is the highest component in Kaavio. All states are handled here and passed down as props to other components.
 *
 * You may pass an onReady(kaavio) function to this. This will be called with the Kaavio reference when everything is
 * rendered. You can access the manipulation API via kaavio.manipulator
 */
export class Kaavio extends React.Component<any, any> {
	manipulator: Manipulator;

	constructor(props) {
		super(props);
		this.state = {
			diagramRef: null,
			highlightedNodes: [], // A stack of the highlighted nodes that is pushed and popped
			hiddenEntities: [], // A stack of hidden entities that is pushed and popped
		};
	}

	private onPanZoomReady(panZoom){
		const {diagramRef} = this.state;
		if(! diagramRef) return;
		this.manipulator = new Manipulator(this, panZoom, diagramRef);

		// Fire the onReady function with a reference to Kaavio
		const {onReady} = this.props;
		onReady(this);
	}

	pushHighlighted = (highlighted: highlightedNode | highlightedNode[]) =>  {
		let toHighlight;
		if(highlighted.constructor !== Array){
			toHighlight = [highlighted];
		}
		else {
			toHighlight = highlighted;
		}

		this.setState((prevState) => {
			// Remove same elements so the entity is highlighted with the newly specified color
			const differences = _.differenceWith(prevState.highlightedNodes, toHighlight, (arrVal, othVal) => {
				return arrVal.node_id == othVal.node_id;
			});
			return { highlightedNodes: differences.concat(toHighlight)}
		});
	};

	popHighlighted = (node_id: string | string[]) => {
		let toRemove;
		if(typeof node_id === 'string'){
			toRemove = [node_id]
		}
		else {
			toRemove = node_id;
		}
		const {highlightedNodes} = this.state;

		// Remove any items from the current highlightedNodes array with the same node_id
		_.pullAllWith(highlightedNodes, toRemove, (arrVal, othVal) => {
			return arrVal.node_id == othVal;
		});
		this.setState({highlightedNodes: highlightedNodes});

		this.setState((prevState) => {
			const removed = _.difference(prevState.highlightedNodes, toRemove);
			return {hiddenEntities: removed}
		});
	};

	resetHighlighted = (exclude?: string[]) => {
		const {highlightedNodes} = this.state;
		let toReset = highlightedNodes.map(highlightedNode => {
			return highlightedNode.node_id;
		});
		if(exclude){
			toReset = _.pullAll(toReset, exclude);
		}
		this.popHighlighted(toReset)
	};

	isHighlighted = (node_id: string) => {
		const {highlightedNodes} = this.state;
		return !!highlightedNodes.find(elem => {
			return elem.node_id === node_id;
		});
	};

	pushHidden = (entity_id: string | string[]) => {
		let toHide;
		if(entity_id.constructor !== Array){
			toHide = [entity_id];
		}
		else {
			toHide = entity_id;
		}

		this.setState((prevState) => {
			// Remove same elements so the entity is highlighted with the newly specified color
			const differences = _.differenceWith(prevState.hiddenEntities, toHide, (arrVal, othVal) => {
				return arrVal == othVal;
			});
			return { hiddenEntities: differences.concat(toHide)}
		});
	};

	popHidden = (entity_id: string | string[]) => {
		let toRemove;
		if(typeof entity_id === 'string'){
			toRemove = [entity_id]
		}
		else {
			toRemove = entity_id;
		}

		this.setState((prevState) => {
			const removed = _.difference(prevState.hiddenEntities, toRemove);
			return {hiddenEntities: removed}
		});
	};

	resetHidden = (exclude?: string[]) => {
		const {hiddenEntities} = this.state;

		let toReset = hiddenEntities;
		if(exclude){
			toReset = _.pullAll(toReset, exclude);
		}
		this.popHidden(toReset)
	};

	isHidden = (entity_id: string) => {
		const {hiddenEntities} = this.state;
		return hiddenEntities.indexOf(entity_id) > -1;
	};

	getEntities = () => {
		return this.props.entities;
	};

	render() {
		const {customStyle, filters, handleClick, entities, name, width, height, edgeDrawers, icons,
			markerDrawers, showPanZoomControls = true} = this.props;
		const {highlightedNodes , hiddenEntities} = this.state;

		const backgroundColor = customStyle.backgroundColor || 'white' ;
		let {about} = this.props;
		about = about || ('kaavio-container-' + new Date().toISOString()).replace(/\W/g, '');

		const entityMap = entities.reduce(function(acc, entity) {
			acc[entity.id] = entity;
			return acc;
		}, {});

		const zIndices = entities
			.sort(function(a: any, b: any) {
				if (a.zIndex > b.zIndex) {
					return 1;
				} else if (a.zIndex < b.zIndex) {
					return -1;
				} else {
					return 0;
				}
			})
			.map((entity) => entity.id);

		// TODO: Don't use refs!
		// Accessing the diagram ref from the state is a little bit of a hack to get panZoom working.
		// Consider refactoring the panZoom to be truly Reactive and not use refs
		return (
			<div id={about} className={`kaavio-container ${ customStyle.containerClass }`}>
				<Diagram
					ref={diagram => !this.state.diagramRef && this.setState({diagramRef: diagram})}
					about={`kaavio-diagram-for-${about}`}
					name={name}
					width={width}
					height={height}
					backgroundColor={backgroundColor}
					edgeDrawers={edgeDrawers}
					entityMap={entityMap}
					filters={filters}
					handleClick={handleClick}
					icons={icons}
					markerDrawers={markerDrawers}
					zIndices={zIndices}
					customStyle={customStyle}
					highlightedNodes={highlightedNodes}
					hiddenEntities={hiddenEntities}
				/>
				<PanZoom diagram={this.state.diagramRef} onReady={panZoom => this.onPanZoomReady(panZoom)}
						 showPanZoomControls={showPanZoomControls} />
			</div>
		)
	}
}
