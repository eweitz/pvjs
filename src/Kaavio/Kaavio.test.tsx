import * as React from 'react';
import { shallow, render } from 'enzyme';
import { Kaavio } from './Kaavio';
import toJson from 'enzyme-to-json';

describe('Kaavio', () => {
    it('should render correctly', () => {
        const comp = shallow(<Kaavio customStyle={{}} entities={[]} />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('Should render with an entity list', () => {
        const entities = [
            {
                height: 12,
                id: 'my-node',
                kaavioType: 'node',
                textContent: 'My node',
                color: 'black',
                backgroundColor: 'white',
                borderWidth: 1,
                width: 1.5,
                x: 0,
                y: 0
            }
        ];
        const comp = shallow(<Kaavio customStyle={{}} entities={entities} />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('Should allow highlighted entities', () => {
        const entities = [
            {
                height: 12,
                id: 'my-node',
                kaavioType: 'node',
                textContent: 'My node',
                color: 'black',
                backgroundColor: 'white',
                borderWidth: 1,
                width: 1.5,
                x: 0,
                y: 0
            }
        ];
        const comp = shallow(<Kaavio customStyle={{}} entities={entities}
                                     highlightedEntities={[{entityId: 'my-node', color: 'red'}]} />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('Should allow hiding entities', () => {
        const entities = [
            {
                height: 12,
                id: 'my-node',
                kaavioType: 'node',
                textContent: 'My node',
                color: 'black',
                backgroundColor: 'white',
                borderWidth: 1,
                width: 1.5,
                x: 0,
                y: 0
            }
        ];
        const comp = shallow(<Kaavio customStyle={{}} entities={entities}
                                     hiddenEntities={['my-node']} />);
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('Should allow zoomed and panned entities', () => {
        const entities = [
            {
                height: 12,
                id: 'my-node',
                kaavioType: 'node',
                textContent: 'My node',
                color: 'black',
                backgroundColor: 'white',
                borderWidth: 1,
                width: 1.5,
                x: 0,
                y: 0
            }
        ];
        const comp = shallow(<Kaavio customStyle={{}} entities={entities}
                                     zoomedEntities={['my-node']} pannedEntities={['my-node']} />);
        expect(toJson(comp)).toMatchSnapshot();
    });
});