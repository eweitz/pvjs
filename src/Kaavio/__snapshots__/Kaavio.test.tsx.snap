// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kaavio Should allow hiding entities 1`] = `
<div
  className="kaavio-container undefined"
  id="kaavio-container"
>
  <Diagram
    about="kaavio-diagram-for-kaavio-container"
    backgroundColor="white"
    customStyle={Object {}}
    entityMap={
      Object {
        "my-node": Object {
          "backgroundColor": "white",
          "borderWidth": 1,
          "color": "black",
          "height": 12,
          "id": "my-node",
          "kaavioType": "node",
          "textContent": "My node",
          "width": 1.5,
          "x": 0,
          "y": 0,
        },
      }
    }
    hiddenEntities={
      Array [
        "my-node",
      ]
    }
    highlightedNodes={Array []}
    zIndices={
      Array [
        "my-node",
      ]
    }
  />
  <PanZoom
    diagram={null}
    onReady={[Function]}
    showPanZoomControls={true}
  />
</div>
`;

exports[`Kaavio Should allow highlighted entities 1`] = `
<div
  className="kaavio-container undefined"
  id="kaavio-container"
>
  <Diagram
    about="kaavio-diagram-for-kaavio-container"
    backgroundColor="white"
    customStyle={Object {}}
    entityMap={
      Object {
        "my-node": Object {
          "backgroundColor": "white",
          "borderWidth": 1,
          "color": "black",
          "height": 12,
          "id": "my-node",
          "kaavioType": "node",
          "textContent": "My node",
          "width": 1.5,
          "x": 0,
          "y": 0,
        },
      }
    }
    highlightedNodes={
      Array [
        Object {
          "color": "red",
          "node_id": "my-node",
        },
      ]
    }
    zIndices={
      Array [
        "my-node",
      ]
    }
  />
  <PanZoom
    diagram={null}
    onReady={[Function]}
    showPanZoomControls={true}
  />
</div>
`;

exports[`Kaavio Should allow zoomed and panned entities 1`] = `
<div
  className="kaavio-container undefined"
  id="kaavio-container"
>
  <Diagram
    about="kaavio-diagram-for-kaavio-container"
    backgroundColor="white"
    customStyle={Object {}}
    entityMap={
      Object {
        "my-node": Object {
          "backgroundColor": "white",
          "borderWidth": 1,
          "color": "black",
          "height": 12,
          "id": "my-node",
          "kaavioType": "node",
          "textContent": "My node",
          "width": 1.5,
          "x": 0,
          "y": 0,
        },
      }
    }
    highlightedNodes={Array []}
    zIndices={
      Array [
        "my-node",
      ]
    }
  />
  <PanZoom
    diagram={null}
    onReady={[Function]}
    pannedEntities={
      Array [
        "my-node",
      ]
    }
    showPanZoomControls={true}
    zoomedEntities={
      Array [
        "my-node",
      ]
    }
  />
</div>
`;

exports[`Kaavio Should render with an entity list 1`] = `
<div
  className="kaavio-container undefined"
  id="kaavio-container"
>
  <Diagram
    about="kaavio-diagram-for-kaavio-container"
    backgroundColor="white"
    customStyle={Object {}}
    entityMap={
      Object {
        "my-node": Object {
          "backgroundColor": "white",
          "borderWidth": 1,
          "color": "black",
          "height": 12,
          "id": "my-node",
          "kaavioType": "node",
          "textContent": "My node",
          "width": 1.5,
          "x": 0,
          "y": 0,
        },
      }
    }
    highlightedNodes={Array []}
    zIndices={
      Array [
        "my-node",
      ]
    }
  />
  <PanZoom
    diagram={null}
    onReady={[Function]}
    showPanZoomControls={true}
  />
</div>
`;

exports[`Kaavio should render correctly 1`] = `
<div
  className="kaavio-container undefined"
  id="kaavio-container"
>
  <Diagram
    about="kaavio-diagram-for-kaavio-container"
    backgroundColor="white"
    customStyle={Object {}}
    entityMap={Object {}}
    highlightedNodes={Array []}
    zIndices={Array []}
  />
  <PanZoom
    diagram={null}
    onReady={[Function]}
    showPanZoomControls={true}
  />
</div>
`;
